Support Adding Application Egg to PySpark

rfrohock opened the pull request:
**Pull Request checklist**

- [x ] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior:** https://github.com/spark-jobserver/spark-jobserver/issues/940



**New behavior:** Support  Adding application egg to Pyspark context


**Other information**:
Didn't want to hardcode path version of examples egg, so did not include new unit test, will add if someone has a convention for adding a reference to the generated egg.  all of the existing tests do not make use of any egg files

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/941)
<!-- Reviewable:end -->

zeitos commented on job-server-python/src/python/sparkjobserver/subprocess.py:
```diff
@@ -98,6 +100,15 @@ def import_class(cls):
                     "or HiveContext but received %s" % repr(context_class), 2)
+
+    egg_path = os.environ.get("EGGPATH", None)
+    if egg_path is not None and sc is not None:
```
is empty a valid path?

Also: don't you want to fail if `sc is None`?

rfrohock replied:
will change the check for the path.  there are two cases JavaSparkContext and CustomContext that don't appear to set the SC, not sure why for the JavaSparkContext, in those cases the egg won't be added.  depending on the job you have submitted that might be ok.  all of the python tests worked without the egg being added before

zeitos commented:
@rfrohock Thanks! I think @noorul will need to review this PR before merging 'cos he did all the python stuff

zeitos commented:
@rfrohock also: can you update the doc?

noorul commented on job-server-extras/src/main/scala/spark/jobserver/python/PythonJob.scala:
```diff
@@ -64,6 +64,7 @@ case class PythonJob[X <: PythonContextLike](eggPath: String,
         Process(
           Seq(sc.pythonExecutable, "-m", "sparkjobserver.subprocess", server.getListeningPort.toString),
           None,
+          "EGGPATH" -> eggPath,
```
How are we going to upload the file to SJS node?

rfrohock replied:
the application EGG file would have already been uploaded using standard job binary API.  this patch ensures the already loaded app path is set in the context of the shelled spark instance.

noorul commented:
@rfrohock Can you update the doc?

rfrohock commented:
@noorul Updated Python Doc, removed obsolete Context Factories, added new SessionContext.  Updated code examples.  Let me know if you need anything else

noorul commented:
@rfrohock Did you forget to push the changes?
